### PR TITLE
write a test for spherical selections tokens using the periodic kdtree

### LINKED ISSUE #1731: write a test for spherical selections tokens using the periodic kdtree
[test_atomselections.py](https://github.com/MDAnalysis/mdanalysis/blob/develop/testsuite/MDAnalysisTests/core/test_atomselections.py) is lacking a test using the `SphericalLayer` and `SphericalZone` distance-based atom selections employing distance searches with `PeriodicKDTree`.

### PR DESCRIPTION
**Pull request by jmborr**
Fixes #1731

**_IN PROGRESS_**

PR Checklist
------------
 - [x] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Comment by richardjgowers** [comment]
This looks like it'll be a good addition.  If we can get the kdtree version of all selections working, we should probably make that the only option.

**Comment by jmborr** [comment]
@richardjgowers I just opened issue #1736 to test cylindrical selections 😄

**Comment by jmborr** [comment]
@orbeckst  @richardjgowers I'm running into a bit of a problem here. `test_atomselections.py` uses trajectory file `adk_dims.dcd` which when loaded results in `dimensions=[0, 0, 0, 90, 90, 90]`. Two distance-based selection taking into account periodic boundary conditions are failing because of this.
Do you know how can one end up with such bizarre `dimensions`?

**Comment by orbeckst** [comment]
The trajectory was simulated without PBC. If you need a trajectory with PBC, use TPR/XTC.

You could fudge the PSF/DCD test by setting the dimensions to some reasonable value before doing the selections. 

Incidentally, failing with undefined box is a good corner case to catch.

**Comment by jmborr** [comment]
@orbeckst So do I understand that if I load a trajectory with no PBC, MDAnalysis will always assign `dimensions=[0, 0, 0, 90, 90, 90]` instead of `dimensions=None` or is this just some bizarre occurrence?

**Comment by jmborr** [comment]
OK, then I'll insert a box checking  to prevent the combination of `dimensions=[0, 0, 0, 90, 90, 90]` and the default `MDAnalysis.core.flags['use_periodic_selections']=True` to end up fooling `DistanceSelection.apply()`

**Comment by richardjgowers** [comment]
Isn't this more a quirk of the topology file saying the box is 0,0,0 rather
than Mda saying it? I'll look into this before we uglify the code with
checks

**Comment by orbeckst** [comment]
The topology says nothing about the box. The trajectory has box information – or not. However...

I checked for our XYZ example, which does not contain box information, and I get `dimensions == array([ 0.,  0.,  0.,  0.,  0.,  0.], dtype=float32)`.

Thus, it is *not true* that we always set the box to `[0, 0, 0, 90, 90, 90]` when we don't have the box. Sorry @jmborr , I was wrong. It seems that we are not handling this case uniformly and have left it to the reader implementation. For example, the DCD reader code initializes lengths to 0 and angles to 90º

https://github.com/MDAnalysis/mdanalysis/blob/eb3678da955b768e71f94a70f7622f1f82345666/package/MDAnalysis/lib/formats/libdcd.pyx#L581

whereas the XYZ reader just uses the base defaults (all 0).

Barring a comprehensive survey, I am _assuming_ that having no box information will at least lead to 
```python
all(dimensions[:3] == 0) == True
```
(or the float equivalent `np.allclose()`)

**Comment by orbeckst** [comment]
(I opened #1738 so that this point is not forgotten.)

**Comment by richardjgowers** [comment]
I think the hack here for dimensions is fine, and then one day in the future we'll have nicer handling that is universal across the package

**Comment by jmborr** [comment]
@orbeckst @richardjgowers I'm having an error with [one of the builds](https://travis-ci.org/MDAnalysis/mdanalysis/jobs/316968496) that seems unrelated to the code changes. Have you ever encountered something similar?

**Comment by orbeckst** [comment]
I've seen this one before. For some reason, on the Mac OSX workers, pytest sometimes fails to create a temp directory; possibly an issue with the xdist plugin for running in parallel (??). Not quite sure. (Or maybe we wrote an incorrect pytest fixture somewhere.)

I restarted https://travis-ci.org/MDAnalysis/mdanalysis/jobs/315961274 , maybe it goes away...

**Comment by orbeckst** [comment]
I restarted the Mac OS X travis job again, but if it fails again, I'd be willing to ignore it.

File: package/MDAnalysis/core/selection.py [hunk 1]
```diff
@@ -247,6 +247,23 @@ def __init__(self):
...
+
+        Parameters
+        ----------
+        dimensions : numpy.ndarray
+            6-item array denoting system size and angles
+
+        Returns
+        -------
+        None or numpy.ndarray
+            Returns argument dimensions if system is periodic, otherwise
```
Comment (orbeckst) [inline package/MDAnalysis/core/selection.py]: This description "if system is periodic" is ambiguous because `self.periodic` is an external choice if periodicity ought to be taken into account. Can you please say more clearly what goes into deciding if `None` or the box is returned? 

Also make clear that we only consider full 3D periodicity as periodic, we don't support periodicity in only some dimensions (or arbitrary space group operators... that would be fun).

File: package/MDAnalysis/core/selection.py [hunk 2]
```diff
@@ -327,9 +345,11 @@ def _apply_KDTree(self, group):
 
     def _apply_distmat(self, group):
         sel = self.sel.apply(group)
-        ref = sel.center_of_geometry().reshape(1, 3).astype(np.float32)
-
-        box = group.dimensions if self.periodic else None
+        periodic = False if self.validate_dimensions(group.dimensions) is None\
```
Comment (orbeckst) [inline package/MDAnalysis/core/selection.py]: shorter
```python
periodic = self.validate_dimensions(group.dimensions) is not None
```

File: package/MDAnalysis/core/selection.py [hunk 3]
```diff
@@ -349,24 +369,27 @@ def __init__(self, parser, tokens):
...
         found_indices = kdtree.get_indices()
-
         return group[found_indices].unique
 
     def _apply_distmat(self, group):
         sel = self.sel.apply(group)
-        ref = sel.center_of_geometry().reshape(1, 3).astype(np.float32)
-
-        box = group.dimensions if self.periodic else None
+        periodic = False if self.validate_dimensions(group.dimensions) is None\
```
Comment (orbeckst) [inline package/MDAnalysis/core/selection.py]: shorter
```python
periodic = self.validate_dimensions(group.dimensions) is not None
```

File: package/MDAnalysis/core/selection.py (hunk 2 above)
Comment (orbeckst) [reply package/MDAnalysis/core/selection.py]: But even better, rearrange your code to avoid calling the same method twice:
```python
box = self.validate_dimensions(group.dimensions)
periodic = box is not None
ref = sel.center_of_geometry(pbc=periodic).reshape(1, 3).astype(np.float32)
```

File: package/MDAnalysis/core/selection.py (hunk 3 above)
Comment (orbeckst) [reply package/MDAnalysis/core/selection.py]: But even better, rearrange your code to avoid calling the same method twice:
```python
box = self.validate_dimensions(group.dimensions)
periodic = box is not None
ref = sel.center_of_geometry(pbc=periodic).reshape(1, 3).astype(np.float32)
```

File: package/MDAnalysis/core/selection.py (hunk 1 above)
Comment (jmborr) [reply package/MDAnalysis/core/selection.py]: Then the wording can be changed "periodic in all three dimensions" and the validation code can be changed to:
```
if self.periodic and all(dimensions[:3]):
    return dimensions
return None
```
What do you think?

Comment (orbeckst) [reply package/MDAnalysis/core/selection.py]: Yes, sounds good.

File: package/MDAnalysis/core/selection.py (hunk 2 above)
Comment (jmborr) [reply package/MDAnalysis/core/selection.py]: :+1:

**Comment by orbeckst** [comment]
Please also update CHANGELOG and docs - probably somewhere under selections.

**Comment by orbeckst** [comment]
Generally, we're ok with accuracy to single (float32) precision, which is about 6-7 decimals. Double is nice to have but I don't have a problem if you decrease the number of decimals for the test. If it passes at 10 decimals then that's still good. 

--
Oliver Beckstein
email: orbeckst@gmail.com

**Comment by jmborr** [comment]
@orbeckst There seem to be a conflict in the [python 3.4 build](https://travis-ci.org/MDAnalysis/mdanalysis/jobs/328848140) because the numpy requirement is incompatible with this python version:

```
UnsatisfiableError: The following specifications were found to be in conflict:
  - numpy 1.14* -> python >=2.7,<2.8.0a0
  - python 3.4*
Use "conda info <package>" to see the dependencies for each package.
```

I guess this will affect all pull requests

**Comment by jmborr** [comment]
@kain88-de If you want, you could take a look to this [Mac OS-X build](https://travis-ci.org/MDAnalysis/mdanalysis/jobs/328848147) flagged as "allowed to fail".

**Comment by kain88-de** [comment]
yeah, that is expected. No idea what is wrong with OSX right now. None of us has the time to look into it currently. Therefore it is set to be allowed to fail.

**Comment by jmborr** [comment]
I travel in the same tight ship, so that's fine with me 😸

**Comment by kain88-de** [comment]
You can restart the python 3.4 test if this PR is merged. https://github.com/conda-forge/numpy-feedstock/pull/76

**Comment by jmborr** [comment]
@kain88-de is the [ conda-forge/numpy-feedstock#76](https://github.com/conda-forge/numpy-feedstock/pull/76) PR on ice ?

**Comment by kain88-de** [comment]
Travis is on ice for the last couple days. They still struggle with OS X builds. We are trying to get those to pass. That will likely still take some time.

**Comment by jmborr** [comment]
@kain88-de I wasn't aware Travis was having such an issue, good to know 👍
